pydantic_settings/settings.py

```diff
@@ -345,3 +326,34 @@ def remove_empty_database_dicts(cls, v):
             return {}
 
         return {key: value for key, value in v.dict().items() if value != {}}
+
+    @root_validator
+    def get_dynamic_defaults(cls, values):
+        """
+        Get dynamic defaults for BASE_DIR, ROOT_URLCONF, and WSGI_APPLICATION.
+        """
+        module = inspect.getmodule(cls)
+        if module and module.__name__.startswith("pydantic_settings."):
+            module = None
+
+        base_dir: Path = values["BASE_DIR"]
+
+        if not base_dir and module:
+            path = Path(inspect.getfile(module)).resolve()
+            if path.stem == "__init__":
+                path = path.parent
+            for part in module.__name__.split("."):
+                path = path.parent
+            values["BASE_DIR"] = path
+
+        base_module = None
+        if module:
+            base_module = module.__name__.split(".", 1)[0]
+        elif base_dir:
+            base_module = base_dir.name
```

Comment:
`base_module` should probably only be calculated from `base_dir` (which should also be set after `values["BASE_DIR"] = path`), eliminating the need of the `if module` tree and ensuring BASE_DIR is the source of truth if provided.